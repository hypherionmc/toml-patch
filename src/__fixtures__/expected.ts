export default {
  type: 'document',
  loc: {
    start: { line: 0, column: 0 },
    end: { line: 34, column: 0 }
  },
  body: [
    {
      type: 'comment',
      loc: {
        start: { line: 1, column: 0 },
        end: { line: 1, column: 26 }
      },
      raw: '# This is a TOML document.'
    },
    {
      type: 'keyvalue',
      loc: {
        start: { line: 3, column: 0 },
        end: { line: 3, column: 22 }
      },
      key: {
        type: 'key',
        loc: {
          start: { line: 3, column: 0 },
          end: { line: 3, column: 5 }
        },
        raw: 'title',
        value: ['title']
      },
      value: {
        type: 'string',
        loc: {
          start: { line: 3, column: 8 },
          end: { line: 3, column: 22 }
        },
        raw: `"TOML Example"`,
        value: 'TOML Example'
      },
      equals: 6,
      comments: null
    },
    {
      type: 'table',
      loc: {
        start: { line: 5, column: 0 },
        end: { line: 7, column: 51 }
      },
      key: {
        type: 'tablekey',
        loc: {
          start: { line: 5, column: 0 },
          end: { line: 5, column: 7 }
        },
        value: {
          type: 'key',
          loc: {
            start: { line: 5, column: 1 },
            end: { line: 5, column: 6 }
          },
          raw: 'owner',
          value: ['owner']
        },
        comment: null
      },
      items: [
        {
          type: 'keyvalue',
          loc: {
            start: { line: 6, column: 0 },
            end: { line: 6, column: 27 }
          },
          key: {
            type: 'key',
            loc: {
              start: { line: 6, column: 0 },
              end: { line: 6, column: 4 }
            },
            raw: 'name',
            value: ['name']
          },
          equals: 5,
          value: {
            type: 'string',
            loc: {
              start: { line: 6, column: 7 },
              end: { line: 6, column: 27 },
              raw: `"Tom Preston-Werner"`,
              value: 'Tom Preston-Werner'
            }
          },
          comments: null
        },
        {
          type: 'keyvalue',
          loc: {
            start: { line: 7, column: 0 },
            end: { line: 7, column: 51 }
          },
          key: {
            type: 'key',
            loc: {
              start: { line: 7, column: 0 },
              end: { line: 7, column: 3 }
            },
            raw: 'dob',
            value: ['dob']
          },
          equals: 4,
          value: {
            type: 'datetime',
            loc: {
              start: { line: 7, column: 6 },
              end: { line: 7, column: 31 }
            },
            raw: '1979-05-27T07:32:00-08:00',
            value: new Date('1979-05-27T07:32:00-08:00')
          },
          comments: [
            {
              type: 'comment',
              loc: {
                start: { line: 7, column: 32 },
                end: { line: 7, column: 51 }
              },
              raw: '# First class dates'
            }
          ]
        }
      ]
    },
    {
      type: 'table',
      loc: {
        start: { line: 9, column: 0 },
        end: { line: 13, column: 14 }
      },
      key: {
        type: 'tablekey',
        loc: {
          start: { line: 9, column: 0 },
          end: { line: 9, column: 10 }
        },
        value: {
          type: 'key',
          loc: {
            start: { line: 9, column: 1 },
            end: { line: 9, column: 9 }
          },
          raw: 'database',
          value: ['database']
        },
        comment: null
      },
      items: [
        {
          type: 'keyvalue',
          loc: {
            start: { line: 10, column: 0 },
            end: { line: 10, column: 22 }
          },
          key: {
            type: 'key',
            loc: {
              start: { line: 10, column: 0 },
              end: { line: 10, column: 6 }
            },
            raw: 'server',
            value: ['server']
          },
          equals: 7,
          value: {
            type: 'string',
            loc: {
              start: { line: 10, column: 10 },
              end: { line: 10, column: 22 }
            },
            raw: `"192.168.1.1"`,
            value: '192.168.1.1'
          },
          comments: null
        },
        {
          type: 'keyvalue',
          loc: {
            start: { line: 11, column: 0 },
            end: { line: 11, column: 28 }
          },
          key: {
            type: 'key',
            loc: {
              start: { line: 11, column: 0 },
              end: { line: 11, column: 5 }
            },
            raw: 'ports',
            value: ['ports']
          },
          equals: 6,
          value: {
            type: 'inlinearray',
            loc: {
              start: { line: 11, column: 8 },
              end: { line: 11, column: 28 }
            },
            items: [
              {
                type: 'inlinearrayitem',
                loc: {
                  start: { line: 11, column: 10 },
                  end: { line: 11, column: 14 }
                },
                item: {
                  type: 'integer',
                  loc: {
                    start: { line: 11, column: 10 },
                    end: { line: 11, column: 14 }
                  },
                  raw: '8001',
                  value: 8001
                },
                comma: true
              },
              {
                type: 'inlinearrayitem',
                loc: {
                  start: { line: 11, column: 16 },
                  end: { line: 11, column: 20 }
                },
                item: {
                  type: 'integer',
                  loc: {
                    start: { line: 11, column: 16 },
                    end: { line: 11, column: 20 }
                  },
                  raw: '8001',
                  value: 8001
                },
                comma: true
              },
              {
                type: 'inlinearrayitem',
                loc: {
                  start: { line: 11, column: 22 },
                  end: { line: 11, column: 26 }
                },
                item: {
                  type: 'integer',
                  loc: {
                    start: { line: 11, column: 22 },
                    end: { line: 11, column: 26 }
                  },
                  raw: '8002',
                  value: 8002
                },
                comma: false
              }
            ]
          },
          comments: null
        },
        {
          type: 'keyvalue',
          loc: {
            start: { line: 12, column: 0 },
            end: { line: 12, column: 21 }
          },
          key: {
            type: 'key',
            loc: {
              start: { line: 12, column: 0 },
              end: { line: 12, column: 14 }
            },
            raw: 'connection_max',
            value: ['connection_max']
          },
          equals: 15,
          value: {
            type: 'integer',
            loc: {
              start: { line: 12, column: 17 },
              end: { line: 12, column: 21 }
            },
            raw: '5000',
            value: 5000
          },
          comments: null
        },
        {
          type: 'keyvalue',
          loc: {
            start: { line: 13, column: 0 },
            end: { line: 13, column: 14 }
          },
          key: {
            type: 'key',
            loc: {
              start: { line: 13, column: 0 },
              end: { line: 13, column: 7 }
            },
            raw: 'enabled',
            value: ['enabled']
          },
          equals: 8,
          value: {
            type: 'boolean',
            loc: {
              start: { line: 13, column: 10 },
              end: { line: 13, column: 14 }
            },
            value: true
          },
          comments: null
        }
      ]
    },
    {
      type: 'table',
      loc: {
        start: { line: 15, column: 0 },
        end: { line: 24, column: 15 }
      },
      key: {
        type: 'tablekey',
        loc: {
          start: { line: 15, column: 0 },
          end: { line: 15, column: 9 }
        },
        value: {
          type: 'key',
          loc: {
            start: { line: 15, column: 1 },
            end: { line: 15, column: 8 }
          },
          raw: 'servers',
          value: ['server']
        },
        comment: null
      },
      items: [
        {
          type: 'comment',
          loc: {
            start: { line: 17, column: 2 },
            end: { line: 17, column: 64 }
          },
          raw: '# Indentation (tabs and/or spaces) is allowed but not required'
        },
        {
          type: 'table',
          loc: {
            start: { line: 18, column: 2 },
            end: { line: 20, column: 15 }
          },
          key: {
            type: 'tablekey',
            loc: {
              start: { line: 18, column: 2 },
              end: { line: 18, column: 17 }
            },
            value: {
              type: 'key',
              loc: {
                start: { line: 18, column: 3 },
                end: { line: 18, column: 16 }
              },
              raw: 'servers.alpha',
              value: ['servers', 'alpha']
            },
            comment: null
          },
          items: [
            {
              type: 'keyvalue',
              loc: {
                start: { line: 19, column: 2 },
                end: { line: 19, column: 17 }
              },
              key: {
                type: 'key',
                loc: {
                  start: { line: 19, column: 2 },
                  end: { line: 19, column: 4 }
                },
                raw: 'ip',
                value: ['ip']
              },
              equals: 5,
              value: {
                type: 'string',
                loc: {
                  start: { line: 19, column: 6 },
                  end: { line: 19, column: 17 }
                },
                raw: `"10.0.0.1"`,
                value: '10.0.0.1'
              },
              comments: null
            },
            {
              type: 'keyvalue',
              loc: {
                start: { line: 20, column: 2 },
                end: { line: 20, column: 15 }
              },
              key: {
                type: 'key',
                loc: {
                  start: { line: 20, column: 2 },
                  end: { line: 20, column: 4 }
                },
                raw: 'dc',
                value: ['dc']
              },
              equals: 5,
              value: {
                type: 'string',
                loc: {
                  start: { line: 20, column: 7 },
                  end: { line: 20, column: 15 }
                },
                raw: `"eqdc10"`,
                value: 'eqdc10'
              },
              comments: null
            }
          ]
        },
        {
          type: 'table',
          loc: {
            start: { line: 22, column: 2 },
            end: { line: 24, column: 15 }
          },
          key: {
            type: 'tablekey',
            loc: {
              start: { line: 22, column: 2 },
              end: { line: 22, column: 16 }
            },
            value: {
              type: 'key',
              loc: {
                start: { line: 22, column: 3 },
                end: { line: 22, column: 15 }
              },
              raw: 'servers.beta',
              value: ['servers', 'beta']
            },
            comment: null
          },
          items: [
            {
              type: 'keyvalue',
              loc: {
                start: { line: 23, column: 2 },
                end: { line: 23, column: 17 }
              },
              key: {
                type: 'key',
                loc: {
                  start: { line: 23, column: 2 },
                  end: { line: 23, column: 4 }
                },
                raw: 'ip',
                value: ['ip']
              },
              equals: 5,
              value: {
                type: 'string',
                loc: {
                  start: { line: 23, column: 6 },
                  end: { line: 23, column: 17 }
                },
                raw: `"10.0.0.2"`,
                value: '10.0.0.2'
              },
              comments: null
            },
            {
              type: 'keyvalue',
              loc: {
                start: { line: 24, column: 2 },
                end: { line: 24, column: 15 }
              },
              key: {
                type: 'key',
                loc: {
                  start: { line: 24, column: 2 },
                  end: { line: 24, column: 4 }
                },
                raw: 'dc',
                value: ['dc']
              },
              equals: 5,
              value: {
                type: 'string',
                loc: {
                  start: { line: 24, column: 7 },
                  end: { line: 24, column: 15 }
                },
                raw: `"eqdc10"`,
                value: 'eqdc10'
              },
              comments: null
            }
          ]
        }
      ]
    },
    {
      type: 'table',
      loc: {
        start: { line: 26, column: 0 },
        end: { line: 33, column: 1 }
      },
      key: {
        type: 'tablekey',
        loc: {
          start: { line: 26, column: 0 },
          end: { line: 26, column: 9 }
        },
        value: {
          type: 'key',
          loc: {
            start: { line: 26, column: 1 },
            end: { line: 26, column: 8 }
          },
          raw: 'clients',
          value: ['clients']
        },
        comment: null
      },
      items: [
        {
          type: 'keyvalue',
          loc: {
            start: { line: 27, column: 0 },
            end: { line: 27, column: 37 }
          },
          key: {
            type: 'key',
            loc: {
              start: { line: 27, column: 0 },
              end: { line: 27, column: 4 }
            },
            raw: 'data',
            value: ['data']
          },
          equals: 5,
          value: {
            type: 'inlinearray',
            loc: {
              start: { line: 27, column: 7 },
              end: { line: 27, column: 37 }
            },
            items: [
              {
                type: 'inlinearrayitem',
                loc: {
                  start: { line: 27, column: 9 },
                  end: { line: 27, column: 27 }
                },
                item: {
                  type: 'inlinearray',
                  loc: {
                    start: { line: 27, column: 9 },
                    end: { line: 27, column: 27 }
                  },
                  items: [
                    {
                      type: 'inlinearrayitem',
                      loc: {
                        start: { line: 27, column: 10 },
                        end: { line: 27, column: 17 }
                      },
                      item: {
                        type: 'string',
                        loc: {
                          start: { line: 27, column: 10 },
                          end: { line: 27, column: 17 }
                        },
                        raw: `"gamma"`,
                        value: 'gamma'
                      },
                      comma: true
                    },
                    {
                      type: 'inlinearrayitem',
                      loc: {
                        start: { line: 27, column: 19 },
                        end: { line: 27, column: 26 }
                      },
                      item: {
                        type: 'string',
                        loc: {
                          start: { line: 27, column: 19 },
                          end: { line: 27, column: 26 }
                        },
                        raw: `"delta"`,
                        value: 'delta'
                      },
                      comma: false
                    }
                  ]
                },
                comma: true
              },
              {
                type: 'inlinearrayitem',
                loc: {
                  start: { line: 27, column: 29 },
                  end: { line: 27, column: 35 }
                },
                item: {
                  type: 'inlinearray',
                  loc: {
                    start: { line: 27, column: 29 },
                    end: { line: 27, column: 35 }
                  },
                  items: [
                    {
                      type: 'inlinearrayitem',
                      loc: {
                        start: { line: 27, column: 30 },
                        end: { line: 27, column: 31 }
                      },
                      item: {
                        type: 'integer',
                        loc: {
                          start: { line: 27, column: 30 },
                          end: { line: 27, column: 31 }
                        },
                        raw: '1',
                        value: 1
                      },
                      comma: true
                    },
                    {
                      type: 'inlinearrayitem',
                      loc: {
                        start: { line: 27, column: 33 },
                        end: { line: 27, column: 34 }
                      },
                      item: {
                        type: 'integer',
                        loc: {
                          start: { line: 27, column: 33 },
                          end: { line: 27, column: 34 }
                        },
                        raw: '2',
                        value: 2
                      },
                      comma: false
                    }
                  ]
                },
                comma: false
              }
            ]
          },
          comments: null
        },
        {
          type: 'comment',
          loc: {
            start: { line: 29, column: 0 },
            end: { line: 29, column: 39 }
          },
          raw: '# Line breaks are OK when inside arrays'
        },
        {
          type: 'keyvalue',
          loc: {
            start: { line: 30, column: 0 },
            end: { line: 33, column: 1 }
          },
          key: {
            type: 'key',
            loc: {
              start: { line: 30, column: 0 },
              end: { line: 30, column: 5 }
            },
            raw: 'hosts',
            value: ['hosts']
          },
          equals: 6,
          value: {
            type: 'inlinearray',
            loc: {
              start: { line: 30, column: 8 },
              end: { line: 33, column: 1 }
            },
            items: [
              {
                type: 'inlinearrayitem',
                loc: {
                  start: { line: 31, column: 2 },
                  end: { line: 31, column: 9 }
                },
                item: {
                  type: 'string',
                  loc: {
                    start: { line: 31, column: 2 },
                    end: { line: 31, column: 9 }
                  },
                  raw: `"alpha"`,
                  value: 'alpha'
                },
                comma: true
              },
              {
                type: 'inlinearrayitem',
                loc: {
                  start: { line: 32, column: 2 },
                  end: { line: 32, column: 9 }
                },
                item: {
                  type: 'string',
                  loc: {
                    start: { line: 32, column: 2 },
                    end: { line: 32, column: 9 }
                  },
                  raw: `"omega"`,
                  value: 'omega'
                },
                comma: true
              }
            ]
          },
          comments: null
        }
      ]
    }
  ]
};
