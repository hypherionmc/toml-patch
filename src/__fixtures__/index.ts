import { join } from 'path';
import { readFileSync } from 'fs';

function read(filename: string): string {
  return readFileSync(join(__dirname, filename), { encoding: 'utf8' });
}

export const example = read('example.toml');
export const fruit = read('fruit.toml');
export const hard_example = read('hard-example.toml');
export const hard_example_unicode = read('hard-example-unicode.toml');
export const kitchen_sink = read('kitchen-sink.toml');
export const spec_01_example = read('0A-spec-01-example-v0.4.0.toml');
