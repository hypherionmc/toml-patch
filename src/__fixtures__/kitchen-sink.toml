# This is a TOML document.

title = "TOML Example"

[values]
string = "string..."
integer = [ 1_234 , 0xdead_beef , 0o01234567 , 0o755 , 0b11010110 ]
float = [ 1_234.567 , -0.01 , 5e+22 , 1E6 , inf , -inf , nan , -nan ]
boolean = true
date = [
  1979-05-27T07:32:00Z,
  1979-05-27T00:32:00-07:00,
  1979-05-27T00:32:00.999999-07:00,
  1979-05-27 07:32:00Z,

]

date.local = [
  1979-05-27T07:32:00,
  1979-05-27, # Local Date
  07:32:00    # Local Time
]

array.nested = [ [ 1, 2 ], ["a", "b", "c"] ]
array.trailing = [
  1,
  2, # this is ok
]

table.dotted = { type.name = "pug" }

# Table
[dog  .  "tater.man"]
type.name = "pug"

# TODO [ j . "Êž" . 'l' ]

# Array Table
[[products]]
name = "Hammer"
sku = 738594937

[[products]]

[[products]]
name = "Nail"
sku = 284758393
color = "gray"
