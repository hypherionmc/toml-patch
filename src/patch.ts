import parseTOML from './parse-toml';
import parseJS from './parse-js';
import toJS from './to-js';
import toTOML from './to-toml';
import { Format } from './format';
import { AST, Node, NodeType, KeyValue } from './ast';
import diff, { Change, ChangeType } from './diff';
import traverse, { findByPath } from './traverse';
import { Location } from './location';
import { Position } from 'estree';

export default function patch(existing: string, updated: any, format?: Format): string {
  const existing_ast = parseTOML(existing);
  const existing_js = toJS(existing_ast);
  const updated_ast = parseJS(updated, format);

  const changes = diff(existing_js, updated);
  const patched_ast = applyChanges(existing_ast, updated_ast, changes);

  return toTOML(patched_ast);
}

function applyChanges(original: AST, updated: AST, changes: Change[]): AST {
  // Potential Changes:
  //
  // Add: Add key-value to object, add item to array
  // Edit: Change in value
  // Remove: Remove key-value from object, remove item from array
  // Move: Move item in array
  // Rename: Rename key in key-value
  //
  // Special consideration, inline comments need to move as-needed

  // Map changes to the nodes that they apply to
  const search: Map<Node, Change[]> = new Map();
  changes.forEach(change => {
    const path =
      change.type === ChangeType.Add || change.type === ChangeType.Remove
        ? change.path.slice(0, -1)
        : change.path;
    const node = findByPath(original, path);

    if (!search.has(node)) {
      search.set(node, []);
    }

    search.get(node)!.push(change);
  });

  const offset: { line: number; column: { [line: number]: number } } = {
    line: 0,
    column: {}
  };
  const shiftPosition = (position: Position): Position => {
    const line = position.line + offset.line;
    return {
      line,
      column: position.column + (offset.column[line] || 0)
    };
  };
  const shiftLocation = (node: Node) => {
    node.loc = { start: shiftPosition(node.loc.start), end: shiftPosition(node.loc.end) };
  };
  const shiftStart = (node: Node) => {
    node.loc.start = shiftPosition(node.loc.start);
  };
  const shiftEnd = (node: Node) => {
    node.loc.end = shiftPosition(node.loc.end);
  };

  traverse(original, {
    // For nodes with children, need to update the start separately from the end
    // in case a child has changed offsets internally
    // and apply changes on exit so that all children have already moved

    [NodeType.Table]: {
      enter: shiftStart,
      exit(node) {
        const changes = search.get(node) || [];
        changes.forEach(change => {
          if (change.type === ChangeType.Add) {
            console.log('add', node, change);
          } else if (change.type === ChangeType.Remove) {
            console.log('remove', node, change);
          } else {
            throw new Error(`Unsupported change "${change.type} on table`);
          }
        });

        shiftEnd(node);
      }
    },
    [NodeType.InlineTable]: {
      enter: shiftStart,
      exit(node) {
        const changes = search.get(node) || [];
        changes.forEach(change => {
          if (change.type === ChangeType.Add) {
            console.log('add', node, change);
          } else if (change.type === ChangeType.Remove) {
            console.log('remove', node, change);
          } else {
            throw new Error(`Unsupported change "${change.type}" on inline-table`);
          }
        });

        shiftEnd(node);
      }
    },

    [NodeType.TableArray]: {
      enter: shiftStart,
      exit(node) {
        const changes = search.get(node) || [];
        changes.forEach(change => {
          if (change.type === ChangeType.Add) {
            console.log('add', node, change);
          } else if (change.type === ChangeType.Remove) {
            console.log('remove', node, change);
          } else if (change.type === ChangeType.Move) {
            console.log('move', node, change);
          } else {
            throw new Error(`Unsupported change "${change.type}" on table-array`);
          }
        });

        shiftEnd(node);
      }
    },
    [NodeType.InlineArray]: {
      enter: shiftStart,
      exit(node) {
        const changes = search.get(node) || [];
        changes.forEach(change => {
          if (change.type === ChangeType.Add) {
            console.log('add', node, change);
          } else if (change.type === ChangeType.Remove) {
            console.log('remove', node, change);
          } else if (change.type === ChangeType.Move) {
            console.log('move', node, change);
          } else {
            throw new Error(`Unsupported change "${change.type}" on inline-array`);
          }
        });

        shiftEnd(node);
      }
    },

    [NodeType.KeyValue]: {
      enter: shiftStart,
      exit(node) {
        const changes = search.get(node) || [];
        changes.forEach(change => {
          if (change.type === ChangeType.Edit) {
            const replacement = findByPath(updated, change.path);
            const original_loc = node.value.loc;

            node.value = (replacement as KeyValue).value;

            const { lines, columns } = getSpan(node.value.loc);
            const start = shiftPosition(original_loc.start);
            node.value.loc = {
              start,
              end: {
                line: start.line + lines,
                column: start.column + columns
              }
            };

            offset.line += node.value.loc.end.line - original_loc.end.line;
            offset.column[node.value.loc.end.line] =
              node.value.loc.end.column - original_loc.end.column;
          } else if (change.type === ChangeType.Rename) {
            console.log('rename', node, change);
          } else {
            throw new Error(`Unsupported change "${change.type}" on key-value`);
          }
        });

        shiftEnd(node);
      }
    },
    [NodeType.InlineArrayItem]: {
      enter: shiftStart,
      exit(node) {
        const changes = search.get(node) || [];
        changes.forEach(change => {
          if (change.type === ChangeType.Edit) {
            console.log('edit', node, change);
          } else {
            throw new Error(`Unsupported change "${change.type}" on inline-array-item`);
          }
        });

        shiftEnd(node);
      }
    },

    [NodeType.Document]: {
      enter: shiftStart,
      exit: shiftEnd
    },
    [NodeType.InlineTableItem]: {
      enter: shiftStart,
      exit: shiftEnd
    },
    [NodeType.TableKey]: {
      enter: shiftStart,
      exit: shiftEnd
    },
    [NodeType.TableArrayKey]: {
      enter: shiftStart,
      exit: shiftEnd
    },

    [NodeType.Key]: shiftLocation,
    [NodeType.String]: shiftLocation,
    [NodeType.Integer]: shiftLocation,
    [NodeType.Float]: shiftLocation,
    [NodeType.Boolean]: shiftLocation,
    [NodeType.DateTime]: shiftLocation,
    [NodeType.Comment]: shiftLocation
  });

  return original;
}

function getSpan(location: Location): { lines: number; columns: number } {
  return {
    lines: location.end.line - location.start.line,
    columns: location.end.column - location.start.column
  };
}
