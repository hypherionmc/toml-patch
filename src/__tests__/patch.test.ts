import patch from '../patch';
import { parse } from '../';
import { example } from '../__fixtures__';

test('it should apply edit to key-value', () => {
  const value = parse(example);
  value.owner.name = 'Tim Hall';

  expect(patch(example, value)).toMatchSnapshot();
});

test('it should add key-value to table', () => {
  const value = parse(example);
  value.owner.handle = 'timhall';

  expect(patch(example, value)).toMatchSnapshot();
});

test('it should add key-value to inline table', () => {
  const value = parse(example);
  value.clients.count.d = 4;

  expect(patch(example, value)).toMatchSnapshot();
});

test('it should add to inline array', () => {
  const value = parse(example);
  value.clients.count.d = 4;

  expect(patch(example, value)).toMatchSnapshot();
  value.database.ports.push(8003);
});

test('it should add to table array', () => {
  // TODO
  // const value = parse(example);
  // value.products.splice(1, 0, { name: 'Screwdriver', sku: 123456 });
  //
  // expect(patch(example, value)).toMatchSnapshot();
});

it('should key-value from table', () => {
  const value = parse(example);
  delete value.database.enabled;

  expect(patch(example, value)).toMatchSnapshot();
});

it('should remove element from inline array', () => {
  const value = parse(example);
  value.database.ports.splice(1, 1);

  expect(patch(example, value)).toMatchSnapshot();
});

it('should move elements in inline array', () => {
  const value = parse(example);
  value.clients.data[1][0] = 2;
  value.clients.data[1][1] = 1;

  expect(patch(example, value)).toMatchSnapshot();
});

it('should rename key-value in table', () => {
  const value = parse(example);
  delete value.products[1].color;
  value.products[1].product_color = 'gray';

  expect(patch(example, value)).toMatchSnapshot();
});
