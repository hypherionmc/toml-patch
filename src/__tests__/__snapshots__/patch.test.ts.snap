// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should add key-value to table 1`] = `
"# This is a TOML document. Boom.

title = \\"TOML Example\\"

[owner]
name = \\"Tom Preston-Werner\\"
organization = \\"GitHub\\"
bio = \\"GitHub Cofounder & CEO\\\\nLikes tater tots and beer.\\"
dob = 1979-05-27T07:32:00Z # First class dates? Why not?
handle = \\"timhall\\"
[database]
server = \\"192.168.1.1\\"
ports = [ 8001, 8001, 8002 ]
connection_max = 5000
enabled = true

[servers]

  # You can indent as you please. Tabs or spaces. TOML don't care.
  [servers.alpha]
  ip = \\"10.0.0.1\\"
  dc = \\"eqdc10\\"

  [servers.beta]
  ip = \\"10.0.0.2\\"
  dc = \\"eqdc10\\"
  country = \\"中国\\" # This should be parsed as UTF-8

[clients]
data = [ [\\"gamma\\", \\"delta\\"], [1, 2] ] # just an update to make sure parsers support it

# Line breaks are OK when inside arrays
hosts = [
  \\"alpha\\",
  \\"omega\\"
]

count = { a = 1, b=2   , c = 3 }

# Products

  [[products]]
  name = \\"Hammer\\"
  sku = 738594937

  [[products]]
  name = \\"Nail\\"
  sku = 284758393
  color = \\"gray\\"
"
`;

exports[`it should apply edit to key-value 1`] = `
"# This is a TOML document. Boom.

title = \\"TOML Example\\"

[owner]
name = \\"Tim Hall\\"
organization = \\"GitHub\\"
bio = \\"GitHub Cofounder & CEO\\\\nLikes tater tots and beer.\\"
dob = 1979-05-27T07:32:00Z # First class dates? Why not?

[database]
server = \\"192.168.1.1\\"
ports = [ 8001, 8001, 8002 ]
connection_max = 5000
enabled = true

[servers]

  # You can indent as you please. Tabs or spaces. TOML don't care.
  [servers.alpha]
  ip = \\"10.0.0.1\\"
  dc = \\"eqdc10\\"

  [servers.beta]
  ip = \\"10.0.0.2\\"
  dc = \\"eqdc10\\"
  country = \\"中国\\" # This should be parsed as UTF-8

[clients]
data = [ [\\"gamma\\", \\"delta\\"], [1, 2] ] # just an update to make sure parsers support it

# Line breaks are OK when inside arrays
hosts = [
  \\"alpha\\",
  \\"omega\\"
]

count = { a = 1, b=2   , c = 3 }

# Products

  [[products]]
  name = \\"Hammer\\"
  sku = 738594937

  [[products]]
  name = \\"Nail\\"
  sku = 284758393
  color = \\"gray\\"
"
`;
